drivers/platform/maxim/maxim_uart.c

```diff
@@ -0,0 +1,323 @@
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc - The UART descriptor.
+ * @param param - The structure that contains the UART parameters.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t uart_init(struct uart_desc **desc, struct uart_init_param *param)
+{
+	if (!param || !desc)
+		return -EINVAL;
+	int ret = 0;
+	uint32_t did = param->device_id;
+
+	struct uart_desc *descriptor = calloc(1, sizeof(*descriptor));
+
+	uart_cfg_t maxim_desc;
+	sys_cfg_uart_t maxim_desc_sys;
+
+	if (!descriptor)
+		return -ENOMEM;
+
+	maxim_desc_sys.map = MAP_A;
+	maxim_desc_sys.flow_flag = UART_FLOW_DISABLE;
+
+	descriptor->baud_rate = param->baud_rate;
+	descriptor->device_id = param->device_id;
+
+	maxim_desc.parity = UART_PARITY_DISABLE;
+
+	switch(param->size) {
+	case UART_CS_5:
+		maxim_desc.size = UART_DATA_SIZE_5_BITS;
+		break;
+	case UART_CS_6:
+		maxim_desc.size = UART_DATA_SIZE_6_BITS;
+		break;
+	case UART_CS_7:
+		maxim_desc.size = UART_DATA_SIZE_7_BITS;
+		break;
+	case UART_CS_8:
+		maxim_desc.size = UART_DATA_SIZE_8_BITS;
+		break;
+	default:
+		ret = -EINVAL;
+		goto error;
+	}
+
+	switch(param->stop) {
+	case UART_STOP_1_BIT:
+		maxim_desc.stop = UART_STOP_1;
+		break;
+	case UART_STOP_2_BIT:
+		maxim_desc.stop = UART_STOP_2;
+		break;
+	default:
+		ret = -EINVAL;
+		goto error;
+	}
+
+	maxim_desc.baud = param->baud_rate;
+	maxim_desc.flow = UART_FLOW_CTRL_EN;
+	maxim_desc.pol = UART_FLOW_POL_EN;
+
+	uint32_t err = UART_Init(MXC_UART_GET_UART(did), &maxim_desc, &maxim_desc_sys);
+	if (err != E_NO_ERROR) {
+		ret = -EIO;
+		goto error;
+	}
+
+	*desc = descriptor;
+	return 0;
+error:
+	free(descriptor);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by uart_init().
+ * @param desc - The UART descriptor.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t uart_remove(struct uart_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+
+	uart_unregister_callback(desc->device_id);
+	free(desc);
+
+	return 0;
+}
+
+/* Check if UART errors occurred. */
+uint32_t uart_get_errors(struct uart_desc *desc)
+{
+	return 0;
+}
+
+/**
+ * @brief Register a function to be called when an interrupt occurs.
+ * @param port - The UART port.
+ * @param desc - The callback descriptor.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t uart_register_callback(uint8_t port, struct callback_desc *desc)
+{
+	if (!desc || port >= N_PORTS)
+		return -EINVAL;
+	if (!cb[port]) {
+		cb[port] = calloc(1, sizeof(*cb[port]));
+		if (!cb[port])
+			return -ENOMEM;
+	}
+
+	cb[port]->ctx = desc->ctx;
+	cb[port]->callback = desc->callback;
+	cb[port]->config = desc->config;
+
+	return 0;
+}
+
+/**
+ * @brief Unregister a callback function.
+ * @param port - The UART port.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t uart_unregister_callback(uint8_t port)
+{
+	if (port >= N_PORTS || !cb[port])
+		return -EINVAL;
+
+	free(cb[port]);
```

Comment:
set the cb[port] to NULL after free, it could be accessed again
https://stackoverflow.com/questions/1025589/setting-variable-to-null-after-free